lib/Conversion/FSMToSV/FSMToSV.cpp

```diff
@@ -0,0 +1,540 @@
+  StateConversionResults stateConvResults;
+  for (auto state : machineOp.getBody().getOps<StateOp>()) {
+    auto stateConvRes = convertState(state);
+    if (failed(stateConvRes)) {
+      bb.abandon();
+      return failure();
+    }
+    stateConvResults[state] = stateConvRes.getValue();
+    orderedStates.push_back(state);
+    nextStateFromState[state] = stateConvRes.getValue().nextState;
+  }
+
+  // Next-state maps for each
+  llvm::SmallVector<llvm::SmallDenseMap<StateOp, Value>, 4> nextStateMaps;
+
+  // Create next-state mux.
+  nextStateMaps.push_back(nextStateFromState);
+
+  // Create output port muxes.
+  for (size_t portIndex = 0; portIndex < machineOp.getNumResults();
+       portIndex++) {
+    auto &nsmap = nextStateMaps.emplace_back();
+    for (auto &state : orderedStates)
+      nsmap[state] = stateConvResults[state].outputs[portIndex];
+  }
+
+  // Materialize the case mux. We do this in a single call to have a single
+  // always_comb block.
+  auto stateCaseMuxes = buildStateCaseMux(
+      machineOp.getLoc(), stateReg, nextStateMaps, [&](size_t idx) {
+        if (idx == 0)
+          return b.getStringAttr("next_state");
+
+        return b.getStringAttr("output_" + std::to_string(idx - 1));
+      });
+
+  nextStateBackedge.setValue(b.create<sv::ReadInOutOp>(loc, stateCaseMuxes[0]));
+
+  llvm::SmallVector<Value> outputPortAssignments;
+  for (auto outputMux : llvm::makeArrayRef(stateCaseMuxes).drop_front())
+    outputPortAssignments.push_back(
+        b.create<sv::ReadInOutOp>(machineOp.getLoc(), outputMux));
+
+  // Delete the default created output op and replace it with the output muxes.
+  auto *oldOutputOp = hwModuleOp.front().getTerminator();
+  b.create<hw::OutputOp>(loc, outputPortAssignments);
+  oldOutputOp->erase();
+
+  // Erase the original machine op.
+  machineOp.erase();
+
+  return success();
+}
+
+FailureOr<Value>
+MachineOpConverter::convertTransitions( // NOLINT(misc-no-recursion)
+    StateOp currentState, ArrayRef<TransitionOp> transitions) {
+  Value nextState;
+  if (transitions.empty()) {
+    // Base case - transition to the current state.
+    nextState = encoding->encode(currentState);
+  } else {
+    // Recursive case - transition to a named state.
+    auto transition = cast<fsm::TransitionOp>(transitions.front());
+    nextState = encoding->encode(transition.getNextState());
+    if (transition.hasGuard()) {
+      // Not always taken; recurse and mux between the targeted next state and
+      // the recursion result, selecting based on the provided guard.
+      auto guardOpRes = moveOps(&transition.guard().front());
+      if (failed(guardOpRes))
+        return failure();
+
+      auto guard = cast<ReturnOp>(*guardOpRes).getOperand(0);
+      auto otherNextState =
+          convertTransitions(currentState, transitions.drop_front());
+      if (failed(otherNextState))
+        return failure();
+      comb::MuxOp nextStateMux = b.create<comb::MuxOp>(
```

Comment:
Creating a tree of muxes seems like it'd create pretty ugly SV. As a future improvement, could we use some SV behavioral construct for this? Like a `sv.casex`? Or is there a pass which detects this and up-levels it?

I think this is totally acceptable for now but please add a TODO if you agree.

Reply:
> is there a pass which detects this and up-levels it

Nope - by up-leveling I assume you mean (since this is a HW construct) transforming an N long priority-encoded sequence of muxes into `hw.array_get/hw.array_create` ops with `N^2` values in the `hw.array_get`? 

> As a future improvement, could we use some SV behavioral construct for this? Like a sv.casex?

I agree that this may not result in the prettiest of IR. However, I'd eventually also like to see better support for these kinds of patterns in HW such that we don't have to resort to SV and forego any optimizations that may apply at the HW level. If users report that the generated IR is hard to debug then i would not be against making this a `sv.casex`.

Reply:
> > is there a pass which detects this and up-levels it
> 
> Nope - by up-leveling I assume you mean (since this is a HW construct) transforming an N long priority-encoded sequence of muxes into `hw.array_get/hw.array_create` ops with `N^2` values in the `hw.array_get`?

No by "up-level" I meant up-level in terms of SV. Create an `sv` casex priority-based mux.

> If users report that the generated IR is hard to debug then i would not be against making this a `sv.casex`.

+1